In [6]:
#test query for smoking/diabetes

import sqlite3

conn=sqlite3.connect("/Users/dc/clinical.sqlite")
c = conn.cursor()

c.execute("select distinct(user_id) from problems where problem_name like '%smoke%';")
print "users who have problem_name like smoke:"
user_prob_smoke_list = c.fetchall()
print user_prob_smoke_list

#
c.execute("select distinct(user_id) from problems where problem_name like '%smoking%';")
print 'users who have problem_name like smoking:'
user_prob_smoking_list=c.fetchall()
print user_prob_smoking_list

#synonym expansion, words next to smoking for all 4 tables
#tobacco (from table problems)
#affecting (from table problems)
#cessation (from table problems)
#OF (from table problems)
#synonym expaneion, words 2 away 
# HX  (from table problems)
# HISTORY  (from table problems)
# PREGNANCY  (from table problems)
# COUNSELING  (from table problems)
#------------------------------------
# 
# sqlite> select * from medications where medication_name like '%smoking%';
#212|Bupropion Hcl Er (Smoking Det) 150 Mg Or Tb12|993518|Bupropion Hcl Er (Smoking Det) 150 Mg Or Tb12|rxnorm|extracted|56fab5216465d212000e32dc
#212|Bupropion Hcl Er (Smoking Det) 150 Mg Or Tb12|65084-414-42|Bupropion Hcl Er (Smoking Det) 150 Mg Or Tb12|ndc|extracted|56fab5216465d212000e32dc
#212|Bupropion Hcl Er (Smoking Det) 150 Mg Or Tb12|84764|Bupropion Hcl Er (Smoking Det) 150 Mg Or Tb12|medispandrugdescriptorid|extracted|56fab5216465d212000e32dc
#212|Bupropion Hcl Er (Smoking Det) 150 Mg Or Tb12|62100002107430|Bupropion Hcl Er (Smoking Det) 150 Mg Or Tb12|medispangenericproductidentifier|extracted|56fab5216465d212000e32dc
#212|Bupropion Hcl Er (Smoking Det) 150 Mg Or Tb12|84764|Bupropion Hcl Er (Smoking Det) 150 Mg Or Tb12|medfile(medispan)|extracted|56fab5216465d212000e32dc
# Bupropion words 3 away
#no results from 2 other tables
#sqlite> select * from allergies where allergy_name like '%smoking%';
#sqlite> select * from immunizations where immunization_name like '%smoking%';

#problem: synonym expansion doesn't pick up nicotine patch!
c.execute("select distinct(user_id) from medications where medication_name like '%nicotine%';")
print 'users who have medication like nicotine'
user_medication_nicotine = c.fetchall()
print user_medication_nicotine


c.execute("select distinct(user_id) from problems where problem_name like '%copd%';")
print 'users who have problem_name like copd:'
user_prob_copd_list = c.fetchall()
print user_prob_copd_list

#tables allergies and immunizaitons had no results as seen above. 

users who have problem_name like smoke:
[(u'212',), (u'2079',), (u'1923',), (u'74',), (u'2044',), (u'397',), (u'931',)]
users who have problem_name like smoking:
[(u'1952',), (u'2142',), (u'359',), (u'1781',)]
users who have medication like nicotine
[(u'212',), (u'2475',), (u'2118',), (u'312',), (u'1914',), (u'2212',), (u'112',)]
users who have problem_name like copd:
[(u'1593',), (u'596',), (u'1609',)]


<h4>There are different codes for smoking</h4>

<p>
1952|SMOKING CESSATION COUNSELING|V65.42|SMOKING CESSATION COUNSELING|icd9cm|extracted|56fb5d8811cb1f120046015d
1952|SMOKING CESSATION COUNSELING|Z71.6|SMOKING CESSATION COUNSELING|icd10cm|extracted|56fb5d8811cb1f120046015d
2142|Tobacco smoking affecting pregnancy|427189007|Tobacco smoking affecting pregnancy|snomedct|extracted|56ec124944de480100c5ad63
2142|Tobacco smoking affecting pregnancy|649.00|Tobacco smoking affecting pregnancy|icd9cm|extracted|56ec124944de480100c5ad63
2142|Tobacco smoking affecting pregnancy|O99.330|Tobacco smoking affecting pregnancy|icd10cm|extracted|56ec124944de480100c5ad63
359|Tobacco smoking affecting pregnancy in first trimester|427189007|Tobacco smoking affecting pregnancy in first trimester|snomedct|extracted|56dca59db1b5490100418931
359|Tobacco smoking affecting pregnancy in first trimester|649.03|Tobacco smoking affecting pregnancy in first trimester|icd9cm|extracted|56dca59db1b5490100418931
359|Tobacco smoking affecting pregnancy in first trimester|O99.331|Tobacco smoking affecting pregnancy in first trimester|icd10cm|extracted|56dca59db1b5490100418931
1781|HISTORY OF SMOKING|V15.82|HISTORY OF SMOKING|icd9cm|extracted|56d51a5f5219770100de807a
1781|HISTORY OF SMOKING|Z87.891|HISTORY OF SMOKING|icd10cm|extracted|56d51a5f5219770100de807a
1781|HISTORY OF SMOKING|392521001|Medical History|snomedct|inferred|56d51a5f5219770100de807a
1781|HX OF SMOKING|V15.82|HX OF SMOKING|icd9cm|extracted|56d51a495219770100de7d64
1781|HX OF SMOKING|Z87.891|HX OF SMOKING|icd10cm|extracted|56d51a495219770100de7d64
</p>

<h4>Tobacco keyword</h4>
Tobacco is not always synonymous with smoking. ie tobacco chewing

sqlite> select * from problems where problem_name like '%tobacco%';
46|TOBACCO USER|110483000|TOBACCO USER|snomedct|extracted|56f764a2f7cea312007ef66c
46|TOBACCO USER|305.1|TOBACCO USER|icd9cm|extracted|56f764a2f7cea312007ef66c
46|TOBACCO USER|Z72.0|TOBACCO USER|icd10cm|extracted|56f764a2f7cea312007ef66c
9|Tobacco abuse|89765005|Tobacco abuse|snomedct|extracted|56f7162d67d7a901000bc9c9
9|Tobacco abuse|305.1|Tobacco abuse|icd9cm|extracted|56f7162d67d7a901000bc9c9
9|Tobacco abuse|Z72.0|Tobacco abuse|icd10cm|extracted|56f7162d67d7a901000bc9c9
2475|Tobacco use|110483000|Tobacco use|snomedct|extracted|56f1f118a6f66a0100e0d36d
2475|Tobacco use|305.1|Tobacco use|icd9cm|extracted|56f1f118a6f66a0100e0d36d
2475|Tobacco use|Z72.0|Tobacco use|icd10cm|extracted|56f1f118a6f66a0100e0d36d
74|TOBACCO SMOKER|77176002|TOBACCO SMOKER|snomedct|extracted|56f0a04265ff3a1200e0c343
74|TOBACCO SMOKER|305.1|TOBACCO SMOKER|icd9cm|extracted|56f0a04265ff3a1200e0c343
74|TOBACCO SMOKER|F17.200|TOBACCO SMOKER|icd10cm|extracted|56f0a04265ff3a1200e0c343
2142|Tobacco smoking affecting pregnancy|427189007|Tobacco smoking affecting pregnancy|snomedct|extracted|56ec124944de480100c5ad63
2142|Tobacco smoking affecting pregnancy|649.00|Tobacco smoking affecting pregnancy|icd9cm|extracted|56ec124944de480100c5ad63
2142|Tobacco smoking affecting pregnancy|O99.330|Tobacco smoking affecting pregnancy|icd10cm|extracted|56ec124944de480100c5ad63
2142|Tobacco use|110483000|Tobacco use|snomedct|extracted|56ec124944de480100c5acca
2142|Tobacco use|305.1|Tobacco use|icd9cm|extracted|56ec124944de480100c5acca
2142|Tobacco use|Z72.0|Tobacco use|icd10cm|extracted|56ec124944de480100c5acca
774|Tobacco abuse|89765005|Tobacco abuse|snomedct|extracted|56e62d06804f6e0100f2fa73
774|Tobacco abuse|305.1|Tobacco abuse|icd9cm|extracted|56e62d06804f6e0100f2fa73
774|Tobacco abuse|Z72.0|Tobacco abuse|icd10cm|extracted|56e62d06804f6e0100f2fa73
1440|Tobacco abuse|89765005|Tobacco abuse|snomedct|extracted|56e2bfa706e7280100f31a13
1440|Tobacco abuse|305.1|Tobacco abuse|icd9cm|extracted|56e2bfa706e7280100f31a13
1440|Tobacco abuse|Z72.0|Tobacco abuse|icd10cm|extracted|56e2bfa706e7280100f31a13
1794|Tobacco use disorder|89765005|Tobacco use disorder|snomedct|extracted|56e26eb33da9a60100843272
1794|Tobacco use disorder|305.1|Tobacco use disorder|icd9cm|extracted|56e26eb33da9a60100843272
1794|Tobacco use disorder|Z72.0|Tobacco use disorder|icd10cm|extracted|56e26eb33da9a60100843272
2153|Tobacco use disorder|89765005|Tobacco use disorder|snomedct|extracted|56e08f0cb81e7c01004d578a
2153|Tobacco use disorder|305.1|Tobacco use disorder|icd9cm|extracted|56e08f0cb81e7c01004d578a
2153|Tobacco use disorder|Z72.0|Tobacco use disorder|icd10cm|extracted|56e08f0cb81e7c01004d578a
397|TOBACCO SMOKER|77176002|TOBACCO SMOKER|snomedct|extracted|56de4a9ec6f3f301009cea66
397|TOBACCO SMOKER|305.1|TOBACCO SMOKER|icd9cm|extracted|56de4a9ec6f3f301009cea66
397|TOBACCO SMOKER|F17.200|TOBACCO SMOKER|icd10cm|extracted|56de4a9ec6f3f301009cea66
248|Tobacco dependence|89765005|Tobacco dependence|snomedct|extracted|56de461cb81e7c01004c2efc
248|Tobacco dependence|305.1|Tobacco dependence|icd9cm|extracted|56de461cb81e7c01004c2efc
248|Tobacco dependence|F17.200|Tobacco dependence|icd10cm|extracted|56de461cb81e7c01004c2efc
2354|Tobacco use|110483000|Tobacco use|snomedct|extracted|56dce39cb81e7c01004b78aa
2354|Tobacco use|305.1|Tobacco use|icd9cm|extracted|56dce39cb81e7c01004b78aa
2354|Tobacco use|Z72.0|Tobacco use|icd10cm|extracted|56dce39cb81e7c01004b78aa
359|Tobacco smoking affecting pregnancy in first trimester|427189007|Tobacco smoking affecting pregnancy in first trimester|snomedct|extracted|56dca59db1b5490100418931
359|Tobacco smoking affecting pregnancy in first trimester|649.03|Tobacco smoking affecting pregnancy in first trimester|icd9cm|extracted|56dca59db1b5490100418931
359|Tobacco smoking affecting pregnancy in first trimester|O99.331|Tobacco smoking affecting pregnancy in first trimester|icd10cm|extracted|56dca59db1b5490100418931
268|Tobacco abuse, in remission|191889006|Tobacco abuse, in remission|snomedct|extracted|56dbc15ac278dd01006a62bf
268|Tobacco abuse, in remission|V15.82|Tobacco abuse, in remission|icd9cm|extracted|56dbc15ac278dd01006a62bf
268|Tobacco abuse, in remission|F17.201|Tobacco abuse, in remission|icd10cm|extracted|56dbc15ac278dd01006a62bf

<h4>Stemming benchmarking</h4>

we had to manually figure out sql queries like "smoke" and like "smoking"


sqlite> select distinct(problem_name) from problems where problem_name like '%smoke%';
<li>Current smoker</li>
<li>Smoker unmotivated to quit</li>
<li>Former smoker</li>
<li>TOBACCO SMOKER</li>
<li>Smoker</li>
sqlite> select distinct(problem_name) from problems where problem_name like '%smoking%';
<li>SMOKING CESSATION COUNSELING</li>
<li>Tobacco smoking affecting pregnancy</li>
<li>Tobacco smoking affecting pregnancy in first trimester</li>
<li>HISTORY OF SMOKING</li>
<li>HX OF SMOKING</li>
sqlite>



they arent the same
can stemming figure this out? 
is this better than going through the problems table and doing a subsring search to unify 
the smoke and smoking terms? 

What are the statistics for n problems/m terms. 
How to treat multiterm problems? how to unify? 

In [3]:
#1) does stem(smoking)==smoke? yes but not for wordnet
from stemming.porter2 import stem
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.lancaster import LancasterStemmer
from nltk.stem import WordNetLemmatizer


lan_stem = LancasterStemmer()
snow_stem = SnowballStemmer('english')
word_lemma = WordNetLemmatizer()



print "lancaster stem(smoke):",lan_stem.stem("smoke")," stem(smoking):",lan_stem.stem("smoking") 
print "snowball stem(smoke):",snow_stem.stem("smoke")," stem(smoking):",snow_stem.stem("smoking") 
print "word net lemmatizer(smoke):",word_lemma.lemmatize("smoke")," stem(smoking):",word_lemma.lemmatize("smoking") 

#what does smoke/smoking look like in the database? 



lancaster stem(smoke): smok  stem(smoking): smok
snowball stem(smoke): smoke  stem(smoking): smoke
word net lemmatizer(smoke): smoke  stem(smoking): smoking


In [28]:
# statistics for all problems 

import sqlite3
from stemming.porter2 import stem
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.lancaster import LancasterStemmer
from nltk.stem import WordNetLemmatizer 


lan_stem = LancasterStemmer()
snow_stem = SnowballStemmer('english')
word_lemma = WordNetLemmatizer()




conn=sqlite3.connect("/Users/dc/clinical.sqlite")
c = conn.cursor()

c.execute("select distinct(problem_name) from problems;")
list_Problems = c.fetchall()
print len(list_Problems)


stopwords=set(['i','of','in','and','for','with','on','to','or','not','as','the'])
word_set=set()
for problems in list_Problems:
    #print "processing problem:", problems
    for words in problems[0].split():
        w=words.lower().strip(',').strip('(').strip(')').strip()
        
        if set([w]).issubset(stopwords)==False:
            #print "     adding word:", w
            word_set.add(w)
print "numwords:", len(word_set)
#print word_set
print '-------------------------------'
while (len(word_set)>1):
    word = word_set.pop()
    for test_stem in word_set:
        if word==lan_stem.stem(test_stem):
            print 'lancaster stemmer equal:', word, ' lan_stem:',test_stem
        elif word==snow_stem.stem(test_stem):
            print 'snowball stemmer equal:', word, 'snow_stem:', test_stem
        elif word==word_lemma.lemmatize(test_stem):
            print "lemma equal:", word, " to:", test_stem
#test stemming 
# 1 to n; n = 2402
print 'asdfasdfasdf'



2813
numwords: 2402
-------------------------------
lancaster stemmer equal: sleep  lan_stem: sleeping
lancaster stemmer equal: anemia  lan_stem: anemias
lancaster stemmer equal: second  lan_stem: secondary
lancaster stemmer equal: neg  lan_stem: negative
lemma equal: miscarriage  to: miscarriages
lancaster stemmer equal: postop  lan_stem: postoperative
lancaster stemmer equal: wart  lan_stem: warts
lemma equal: metatarsal  to: metatarsals
lancaster stemmer equal: dizzy  lan_stem: dizziness
lancaster stemmer equal: lac  lan_stem: laceration
snowball stemmer equal: hashimoto snow_stem: hashimoto's
lancaster stemmer equal: hashimoto  lan_stem: hashimotos
snowball stemmer equal: over snow_stem: overeating
snowball stemmer equal: bone snow_stem: bones
lancaster stemmer equal: tonsil  lan_stem: tonsils
snowball stemmer equal: chronic snow_stem: chronicity
lemma equal: calorie  to: calories
lemma equal: palpitation  to: palpitations
lancaster stemmer equal: palsy  lan_stem: palsies
lancaster

In [22]:
import sqlite3

conn=sqlite3.connect("/Users/dc/clinical.sqlite")
c = conn.cursor()
from nltk.stem.snowball import SnowballStemmer
from nltk.stem.lancaster import LancasterStemmer
from nltk.stem import WordNetLemmatizer


lan_stem = LancasterStemmer()
snow_stem = SnowballStemmer('english')
word_lemma = WordNetLemmatizer()


def ls(wordstring):
    word_list = wordstring.split()
    stemmed=[]
    for word in word_list:
        stemmed.append(lan_stem.stem(word))
    return stemmed

def ss(wordstring):
    word_list = wordstring.split()
    stemmed=[]
    for word in word_list:
        stemmed.append(snow_stem.stem(word))
    return stemmed

def wl(wordstring):
    word_list = wordstring.split()
    stemmed=[]
    for word in word_list:
        stemmed.append(word_lemma.lemmatize(word))
    return stemmed

c.execute("select distinct(problem_name) from problems;")
list_Problems = c.fetchall()
print len(list_Problems)

# stem phrases has no effect
num=0
for problem in list_Problems:
    if (num%100) ==0:
        print num,": problem",problem, "snowballStem:", snow_stem.stem(problem[0]), " lancasterStem:", lan_stem.stem(problem[0]), " lemmatizer:", word_lemma.lemmatize(problem[0])
    num +=1
print 'end test'

        

    
print '----------------------------------------------'
#stem each word in a phrase
num=0
for problem in list_Problems:
    if num%100 ==0:
        print num,": problem:", problem, "snowballStem:", snow_stem.stem(problem[0]), "lancasterStemmer:", lan_stem.stem(problem[0]), "lemmatizer:", word_lemma.lemmatize(problem[0]) 
    num+=1
    
    
        
        
        

2813
0 : problem (u'Disease',) snowballStem: diseas  lancasterStem: diseas  lemmatizer: Disease
100 : problem (u'BACK PAIN',) snowballStem: back pain  lancasterStem: back pain  lemmatizer: BACK PAIN
200 : problem (u'Oligomenorrhea (finding)',) snowballStem: oligomenorrhea (finding)  lancasterStem: oligomenorrhea (finding)  lemmatizer: Oligomenorrhea (finding)
300 : problem (u'Ehlers-Danlos syndrome',) snowballStem: ehlers-danlos syndrom  lancasterStem: ehlers-danlos syndrome  lemmatizer: Ehlers-Danlos syndrome
400 : problem (u'Sleep disturbance, unspecified',) snowballStem: sleep disturbance, unspecifi  lancasterStem: sleep disturbance, unspecified  lemmatizer: Sleep disturbance, unspecified
500 : problem (u'Anxiety problem',) snowballStem: anxiety problem  lancasterStem: anxiety problem  lemmatizer: Anxiety problem
600 : problem (u'MACROMASTIA',) snowballStem: macromastia  lancasterStem: macromast  lemmatizer: MACROMASTIA
700 : problem (u'SVT (supraventricular tachycardia) (HC code)',

In [1]:
import sqlite3

conn=sqlite3.connect("/Users/dc/clinical.sqlite")
c = conn.cursor()

c.execute("select distinct(problem_name) from problems")
prob_list = c.fetchall()
for problems in prob_list:
    print problems

    
#find the user_ids for each problem


(u'Disease',)
(u'Obesity, Class I, BMI 30-34.9',)
(u'HISTORY OF GESTATIONAL DIABETES',)
(u'CALCULUS OF KIDNEY (KIDNEY STONE)',)
(u'HISTORY OF PREECLAMPSIA (HIGH BLOOD PRESSURE AND PROTEIN IN URINE)',)
(u'MAJOR DEPRESSIVE DISORDER, RECURRENT EPISODE, IN PARTIAL REMISSION',)
(u'HX OF GESTATIONAL DM',)
(u'RENAL CALCULUS',)
(u'OBESITY, BMI 35-39.9, ADULT',)
(u'HX OF PREECLAMPSIA',)
(u'HEALTHY ADULT',)
(u'Encounter for screening colonoscopy',)
(u'Chronic kidney disease',)
(u'Pure hyperglyceridemia',)
(u'ASCUS with positive high risk HPV - cin 1 on colpo - repap/annual in 6 months',)
(u'UTI (lower urinary tract infection)',)
(u'Acute sinusitis, unspecified',)
(u'Pain in joint of right knee-medial-dos 8/29/12',)
(u'Chronic cholecystitis with calculus',)
(u'Obesity',)
(u'ADD (attention deficit disorder)',)
(u'Acute maxillary sinusitis',)
(u'Sore throat',)
(u'Exposure to hepatitis C',)
(u'Migraine',)
(u'Cellulitis and abscess of toe, unspecified',)
(u'Ingrowing nail',)
(u'Allergic rhinitis due 

In [2]:
c.execute("select distinct(user_id) from problems")
u_list = c.fetchall()
for u in u_list:
    print u


(u'2037',)
(u'1486',)
(u'776',)
(u'579',)
(u'1942',)
(u'1941',)
(u'50',)
(u'107',)
(u'121',)
(u'2173',)
(u'1651',)
(u'952',)
(u'491',)
(u'1952',)
(u'365',)
(u'482',)
(u'254',)
(u'458',)
(u'1293',)
(u'1572',)
(u'1719',)
(u'14',)
(u'1601',)
(u'135',)
(u'280',)
(u'490',)
(u'224',)
(u'714',)
(u'935',)
(u'1758',)
(u'103',)
(u'655',)
(u'1217',)
(u'910',)
(u'1199',)
(u'402',)
(u'2307',)
(u'212',)
(u'370',)
(u'115',)
(u'540',)
(u'1704',)
(u'2215',)
(u'687',)
(u'2432',)
(u'560',)
(u'1236',)
(u'1478',)
(u'1615',)
(u'1669',)
(u'848',)
(u'2305',)
(u'1457',)
(u'368',)
(u'2150',)
(u'9',)
(u'936',)
(u'2116',)
(u'2069',)
(u'2359',)
(u'1033',)
(u'1046',)
(u'1366',)
(u'2113',)
(u'2376',)
(u'1456',)
(u'1678',)
(u'111',)
(u'1140',)
(u'46',)
(u'1600',)
(u'2127',)
(u'948',)
(u'801',)
(u'2237',)
(u'1876',)
(u'253',)
(u'310',)
(u'1121',)
(u'440',)
(u'2456',)
(u'1975',)
(u'1116',)
(u'264',)
(u'2171',)
(u'1929',)
(u'1288',)
(u'1544',)
(u'1972',)
(u'1727',)
(u'150',)
(u'917',)
(u'509',)
(u'379',)
(u'1593',)
(u'3